### PR TITLE
[RLlib] Don't error out when using agent collector in inference directly after instantiation

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
Signed-off-by: Artur Niederfahrenhorst <artur@anyscale.com>

## Why are these changes needed?

In order to use our policies in inference, we should be able to directly call policy.compute_actions...(), for example after restoring from a checkpoint. This PR makes it so that agent collector's build-for-inference method can be called immediately with a single prev reward, which is what we need for recurrent policies.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/evaluation/collectors/agent_collector.py [hunk 1]
```diff
@@ -493,7 +493,12 @@ def _get_sample_batch(self, batch_data: Dict[str, TensorType]) -> SampleBatch:
     def _cache_in_np(self, cache_dict: Dict[str, List[np.ndarray]], key: str) -> None:
         """Caches the numpy version of the key in the buffer dict."""
         if key not in cache_dict:
-            cache_dict[key] = [_to_float_np_array(d) for d in self.buffers[key]]
+            arrays_to_cache = []
+            for d in self.buffers[key]:
+                # Don't attempt to cache empty arrays
+                if len(d):
+                    arrays_to_cache.append(_to_float_np_array(d))
```
Comment (ArturNiederfahrenhorst) [inline rllib/evaluation/collectors/agent_collector.py:500]: Calling _to_float_np_array with when caching an empty list would throw an index error.

File: rllib/evaluation/tests/test_agent_collector.py [hunk 2]
```diff
@@ -332,6 +332,19 @@ def test_view_requirement_with_shfit_step(self):
         # check if the padding in the beginning is correct
         check(sample_batch["prev_obses"][0], np.ones((3, 1)) * expected_obses[0])
 
+    def test_begin_with_inference(self):
+        # Test if AC is able to build for inference immediately
+        obs_space = gym.spaces.Box(-np.ones(4), np.ones(4))
+        view_reqs = {
+            SampleBatch.T: ViewRequirement(SampleBatch.T),
+            SampleBatch.REWARDS: ViewRequirement("reward", space=obs_space),
```
Comment (kouroshHakha) [inline rllib/evaluation/tests/test_agent_collector.py:340]: space of reward should just be 1 dimensional? why is it called obs_space?

File: rllib/evaluation/tests/test_agent_collector.py [hunk 3]
```diff
@@ -332,6 +332,19 @@ def test_view_requirement_with_shfit_step(self):
...
+            "prev_reward": ViewRequirement("reward", shift=-1),
+        }
+
+        ac = AgentCollector(view_reqs=view_reqs, is_policy_recurrent=True)
```
Comment (kouroshHakha) [inline rllib/evaluation/tests/test_agent_collector.py:344]: wait this test not using AgentCollector properly. If there is no data added to the buffer (neither with `ac.add_init_obs` nor `ac.add_action_reward_next_obs`) how is it suppose to build a batch for inference. I agree that the error is not informative right now (IndexError) but I don't think you should be allowed to call `ac.build_for_inference()` without having anything in the AgentCollector buffers.

**Review by kouroshHakha — COMMENTED** [review]
I have a very important question man.

File: rllib/evaluation/tests/test_agent_collector.py (hunk 3 above)
Comment (kouroshHakha) [reply rllib/evaluation/tests/test_agent_collector.py:344]: for example look at `_simulate_env_steps` on how timestep infos are added to the collector.

File: rllib/evaluation/tests/test_agent_collector.py (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/evaluation/tests/test_agent_collector.py]: My mistake, changed. Thanks

File: rllib/evaluation/tests/test_agent_collector.py (hunk 3 above)
Comment (ArturNiederfahrenhorst) [reply rllib/evaluation/tests/test_agent_collector.py:343]: Makes sense. The problem surfaces when running PPO with LSTM. I just checked and add_init_obs is called before the call that errors out.
I'll come up with a unittest that is essentially a repro of what happens there and we can see again if it's intended usage of the agent collector.

**Comment by ArturNiederfahrenhorst** [comment]
This only fixed a symptom, but not the issue -> actual fix to be found here: #30024

**Pull request closed without merging**